<a href="https://colab.research.google.com/github/pallavibekal/CodeRepository/blob/main/NFL_Analytics.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### NFL Data Analytics 


1.   List item
2.   List item



The data provided captures details of each season from 1999 - 2013. The data captures statistics like - No of games played/started, Touchdowns and details of each player's college and conference stats with their draft rounds. 

More details can be found here - [NFL Drive](https://drive.google.com/drive/folders/1ahmzPsRpfT6LErAwMjpOYK4TjKqwrS1E?usp=sharing_eil_m&ts=61261088)

In [ ]:
import pandas as pd
import matplotlib as plt
import numpy as np

### Upload the data


In [ ]:
df_init = pd.read_csv('NFL_Player_Stats_1999_2013.csv',engine='python')
df_init

Year               Player  Age  ... Broad Jump�(in) Shuttle 3Cone
0     2006  LaDainian Tomlinson   27  ...             0.0    4.21  6.84
1     2013       Peyton Manning   37  ...             NaN     NaN   NaN
2     2007            Tom Brady   30  ...            99.0    4.38  7.20
3     2011        Aaron Rodgers   28  ...           110.0    0.00  7.39
4     2011           Drew Brees   32  ...           105.0    4.21  7.09
...    ...                  ...  ...  ...             ...     ...   ...
8193  2013        Bryan Walters   26  ...             NaN     NaN   NaN
8194  2013     Kerwynn Williams   22  ...           118.0    4.15  7.15
8195  2013      Leon Washington   31  ...           105.0    4.37  6.96
8196  2013         Matt Veldman   25  ...             NaN     NaN   NaN
8197  2013       Ryan Schraeder   25  ...             NaN     NaN   NaN

[8198 rows x 39 columns]

### Explore the data

Find key characteristics of the data
Hint: Use info(), describe()

In [ ]:
df_init.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8198 entries, 0 to 8197
Data columns (total 39 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             8198 non-null   int64  
 1   Player           8198 non-null   object 
 2   Age              8198 non-null   int64  
 3   Hometown         8198 non-null   object 
 4   Home State       7972 non-null   object 
 5   Tm               8198 non-null   object 
 6   G                8198 non-null   int64  
 7   GS               8198 non-null   int64  
 8   Cmp              8197 non-null   float64
 9   Att              8197 non-null   float64
 10  Yds              8197 non-null   float64
 11  TD               8197 non-null   float64
 12  Int              8197 non-null   float64
 13  Att.1            8197 non-null   float64
 14  Yds.1            8197 non-null   float64
 15  Y/A              4212 non-null   float64
 16  TD.1             8197 non-null   float64
 17  Rec           

### Check the shape of the dataframe
Hint: No of rows, colums, missing and unique values

In [ ]:
print ("Rows     : " , df_init.shape[0])
print ("Columns  : " , df_init.shape[1])
print ("\nFeatures : \n" , df_init.columns.tolist())
print ("\nMissing values :  ", df_init.isnull().sum().values.sum())
print ("\nUnique values :  \n",df_init.nunique())

Rows     :  8198
Columns  :  39

Features : 
 ['Year', 'Player', 'Age', 'Hometown', 'Home State', 'Tm', 'G', 'GS', 'Cmp', 'Att', 'Yds', 'TD', 'Int', 'Att.1', 'Yds.1', 'Y/A', 'TD.1', 'Rec', 'Yds.2', 'Y/R', 'TD.2', 'FantPos', 'FantPt', 'Height (inches)', 'Weight', 'College', 'Conference', 'College wins', 'College losses', 'DOB', 'Draft Round', 'Draft Year', 'Wonderlic', '40�Yard', 'Bench Press', 'Vert Leap�(in)', 'Broad Jump�(in)', 'Shuttle', '3Cone']

Missing values :   39624

Unique values :  
 Year                 15
Player             2302
Age                  24
Hometown            952
Home State           66
Tm                   36
G                    17
GS                   17
Cmp                 349
Att                 479
Yds                 899
TD                   45
Int                  29
Att.1               345
Yds.1               905
Y/A                 768
TD.1                 23
Rec                 120
Yds.2              1134
Y/R                1390
TD.2                

In [ ]:
df_init.describe()

Year          Age  ...      Shuttle        3Cone
count  8198.000000  8198.000000  ...  4085.000000  4085.000000
mean   2006.132959    26.650525  ...     3.121625     5.192472
std       4.324761     3.480960  ...     1.892381     3.171802
min    1999.000000    21.000000  ...     0.000000     0.000000
25%    2002.000000    24.000000  ...     0.000000     0.000000
50%    2006.000000    26.000000  ...     4.160000     6.960000
75%    2010.000000    29.000000  ...     4.310000     7.200000
max    2013.000000    44.000000  ...     5.010000     8.380000

[8 rows x 31 columns]

In [ ]:
df_init.isnull().sum()

Year                  0
Player                0
Age                   0
Hometown              0
Home State          226
Tm                    0
G                     0
GS                    0
Cmp                   1
Att                   1
Yds                   1
TD                    1
Int                   1
Att.1                 1
Yds.1                 1
Y/A                3986
TD.1                  1
Rec                   1
Yds.2                 1
Y/R                2260
TD.2                  1
FantPos               0
FantPt             1200
Height (inches)       0
Weight                0
College              20
Conference          376
College wins       1377
College losses     1377
DOB                   0
Draft Round           0
Draft Year            0
Wonderlic          4113
40�Yard            4113
Bench Press        4113
Vert Leap�(in)     4113
Broad Jump�(in)    4113
Shuttle            4113
3Cone              4113
dtype: int64

In [ ]:
df_init[df_init.columns[df_init.isna().any()]].columns


Index(['Home State', 'Cmp', 'Att', 'Yds', 'TD', 'Int', 'Att.1', 'Yds.1', 'Y/A',
       'TD.1', 'Rec', 'Yds.2', 'Y/R', 'TD.2', 'FantPt', 'College',
       'Conference', 'College wins', 'College losses', 'Wonderlic', '40�Yard',
       'Bench Press', 'Vert Leap�(in)', 'Broad Jump�(in)', 'Shuttle', '3Cone'],
      dtype='object')

In [ ]:
df_init[df_init.columns[df_init.isna().any()]]

### Handle Missing values

> Columns with missing values
- homestate - 226 records
- Player Metrics - Cmp, Att, Yds, TD, Int, Att.1, Yds.1, Y/A, TD.1, Rec, YDS.2
, Y/R, TD.2, Fantpt
- College - 73 records (0-53,N/A-20)
- Conference (0-1, N/A -375)
- College Wins & College Losses - 1377 
- Player college metrics before draft pick up - Wonderlic, 40 Yard, Bench Press, Vertical Leap, Broad Jump, Shuttle, 3 Cone
 



In [ ]:
df_init['Y/R'].replace(np.nan,0,inplace=True)

In [ ]:
df_init['Y/A'].replace(np.nan,0,inplace=True)

In [ ]:
#df_Nullrows = df_init[df_init.isnull().any(axis=1)]
#df_Nullrows.shape  

In [ ]:
#df_init['Year'][df_init['Home State'].isnull()].count()

In [ ]:
#df_Nullrows.to_csv('NFL_Data_Null_Rows.csv')

### Explore values in each column except game and draft statistics

In [ ]:
df_init['Year'].value_counts()

2013    594
2012    576
2006    570
2011    564
2003    557
2010    557
2004    550
2007    547
2005    547
2008    536
2002    533
1999    522
2000    522
2009    520
2001    503
Name: Year, dtype: int64

In [ ]:
print('No of Null records for Hometown -',df_init['Hometown'].isnull().sum())
df_hometown = df_init['Hometown'].value_counts()
df_hometown.head(20)

No of Null records for Hometown - 0


0               226
Houston         143
Miami           143
New Orleans     136
Los Angeles     132
Washington       95
Chicago          94
Columbus         83
Jacksonville     79
San Diego        77
Dallas           76
Cleveland        73
Detroit          71
Richmond         56
Tampa            56
Baton Rouge      55
Atlanta          55
Pittsburgh       54
Philadelphia     53
St. Louis        49
Name: Hometown, dtype: int64

In [ ]:
print('No of Null records for HomeState -',df_init['Home State'].isnull().sum())
df_homestate = df_init['Home State'].value_counts()
df_homestate.head(20)

No of Null records for HomeState - 226


CA    1081
FL     746
TX     717
LA     410
OH     376
PA     371
GA     297
VA     296
NJ     269
NC     247
IL     244
MI     237
AL     193
WA     182
NY     180
MS     175
SC     161
MO     116
IN     116
IA     112
Name: Home State, dtype: int64

In [ ]:
df_init['Tm'].value_counts()


TAM    279
GNB    272
ARI    268
DEN    267
OAK    265
JAX    264
ATL    262
DET    262
STL    261
NOR    258
PIT    257
CIN    256
NWE    256
TEN    254
IND    254
NYG    254
BUF    253
CHI    253
MIN    252
BAL    252
NYJ    251
SFO    251
PHI    251
MIA    250
WAS    250
CAR    250
SEA    249
DAL    249
CLE    246
KAN    242
SDG    234
HOU    194
2TM     75
3TM      5
5TM      1
4TM      1
Name: Tm, dtype: int64

In [ ]:
df_init['FantPos'].value_counts()


WR    2758
RB    2503
TE    1747
QB    1190
Name: FantPos, dtype: int64

In [ ]:
df_init['Height (inches)'].value_counts()

73    2165
75    1040
74     975
76     942
71     722
77     710
70     604
69     304
78     229
12     223
68     122
79      74
67      58
80      19
66       7
81       3
65       1
Name: Height (inches), dtype: int64

In [ ]:
df_init['Weight'].value_counts()

0      535
215    287
200    280
210    279
250    277
      ... 
312      1
169      1
306      1
162      1
323      1
Name: Weight, Length: 140, dtype: int64

In [ ]:
print('No of Null records for College -',df_init['College'].isnull().sum())
df_init['College'].value_counts()

No of Null records for College - 20


Miami (FL)           214
Michigan             197
Florida              184
Ohio St.             166
USC                  157
                    ... 
West Georgia           1
Vanden                 1
Minnesota Vikings      1
Tarboro                1
Vashon                 1
Name: College, Length: 352, dtype: int64

In [ ]:
print('No of Null records for Conference -',df_init['Conference'].isnull().sum())
df_init['Conference'].value_counts()

No of Null records for Conference - 376


Big Ten                                              1255
SEC                                                  1244
ACC                                                  1197
Pac 12                                               1126
Big 12                                                526
Mountain West                                         421
American                                              367
Conference USA                                        216
MAC                                                   216
Independent                                           211
Sun Belt                                              126
Big Sky                                               108
Division II                                           106
Missouri Valley                                        95
Colonial Athletic Association                          92
SWAC                                                   79
Ivy                                                    59
Southern      

In [ ]:
print('No of Null records for Date Of Birth -',df_init['DOB'].isnull().sum())
df_init['DOB'].value_counts()

No of Null records for Date Of Birth - 0


00-01-1900    276
24-03-1976     30
18-07-1900     25
08-02-1979     23
12-06-1979     21
             ... 
05-01-1976      1
16-04-1980      1
06-12-1984      1
16-11-1977      1
30-09-1989      1
Name: DOB, Length: 1970, dtype: int64

### Explore the data for player game statistics 
The columns used for this are
- Year [All the years played by player]
- Player [Player Name]
- Age [Age of Player at that year]
- Tm [Team the player played in that year. *Possibilities of this field - Player played for multiple teams in the same year, player played for different teams in different years, player played for same team for all years*]
- G, GS, Cmp, Att, Yds, TD, Int, Att.1, Yds.1, Y/A, TD.1, Rec, YDS.2
, Y/R, TD.2, Fantpt
- FantPos [Data is divided into 4 positions - QB - Quarterback, RB - Running Back, TE - Tight End, WR - Wide Receiver

In [ ]:
df_init.groupby(['Player']).head(10)

Year               Player  Age  ... Broad Jump�(in) Shuttle 3Cone
0     2006  LaDainian Tomlinson   27  ...             0.0    4.21  6.84
1     2013       Peyton Manning   37  ...             NaN     NaN   NaN
2     2007            Tom Brady   30  ...            99.0    4.38  7.20
3     2011        Aaron Rodgers   28  ...           110.0    0.00  7.39
4     2011           Drew Brees   32  ...           105.0    4.21  7.09
...    ...                  ...  ...  ...             ...     ...   ...
8193  2013        Bryan Walters   26  ...             NaN     NaN   NaN
8194  2013     Kerwynn Williams   22  ...           118.0    4.15  7.15
8195  2013      Leon Washington   31  ...           105.0    4.37  6.96
8196  2013         Matt Veldman   25  ...             NaN     NaN   NaN
8197  2013       Ryan Schraeder   25  ...             NaN     NaN   NaN

[8061 rows x 39 columns]

In [ ]:
categoricals = df_init.select_dtypes(exclude=[np.number])
categoricals.describe()

Player Hometown Home State  ...     College Conference         DOB
count          8198     8198       7972  ...        8178       7822        8198
unique         2302      952         66  ...         352         42        1970
top     Steve Smith        0         CA  ...  Miami (FL)    Big Ten  00-01-1900
freq             18      226       1081  ...         214       1255         276

[4 rows x 8 columns]

In [ ]:
df_init.columns

Index(['Year', 'Player', 'Age', 'Hometown', 'Home State', 'Tm', 'G', 'GS',
       'Cmp', 'Att', 'Yds', 'TD', 'Int', 'Att.1', 'Yds.1', 'Y/A', 'TD.1',
       'Rec', 'Yds.2', 'Y/R', 'TD.2', 'FantPos', 'FantPt', 'Height (inches)',
       'Weight', 'College', 'Conference', 'College wins', 'College losses',
       'DOB', 'Draft Round', 'Draft Year', 'Wonderlic', '40�Yard',
       'Bench Press', 'Vert Leap�(in)', 'Broad Jump�(in)', 'Shuttle', '3Cone'],
      dtype='object')

In [ ]:
df_player_game = df_init.iloc[:,:25]
df_player_game.head()

NameError: ignored

In [ ]:
main_variables = df_player_game.groupby('Player', as_index=False).agg({'Year': 'count', 'Age': 'max','Tm':'count','FantPt':'sum','G':'sum','GS':'sum','Cmp':'sum','Att':'sum','Yds':'sum','TD':'sum','Int':'sum','Att.1':'sum','Yds.1':'sum','Y/A':'sum','TD.1':'sum','Rec':'sum','Yds.2':'sum','Y/R':'sum','TD.2':'sum','Height (inches)':'mean','Weight':'mean'})
main_variables.sort_values(by='FantPt')
main_variables


Player  Year  Age  Tm  ...    Y/R  TD.2  Height (inches)      Weight
0   A.J. Feeley     6   34   6  ...   0.00   0.0             75.0  187.500000
1    A.J. Green     3   25   3  ...  44.73  29.0             76.0  205.000000
2  A.J. Jenkins     2   24   2  ...  16.25   0.0             73.0  192.000000
3  A.J. Ofodile     1   26   1  ...   6.25   0.0             78.0  260.000000
4  Aaron Brooks     7   30   7  ...   1.00   0.0             76.0  125.714286

[5 rows x 22 columns]

In [ ]:
main_variables.sort_values(by='FantPt',ascending=False).head(20)



Player  Year  Age  ...   TD.2  Height (inches)      Weight
1750       Peyton Manning    14   37  ...    0.0        77.000000  115.000000
2143            Tom Brady    14   36  ...    0.0        76.000000  112.500000
706            Drew Brees    13   34  ...    1.0        73.000000   96.461538
1367  LaDainian Tomlinson    11   32  ...   17.0        59.454545  221.000000
690        Donovan McNabb    13   35  ...    0.0        74.000000  129.230769
229           Brett Favre    12   41  ...    0.0        74.000000  112.500000
1790           Randy Moss    13   35  ...  139.0        76.000000  215.000000
144    Ben Roethlisberger    10   31  ...    0.0        77.000000   96.000000
2092        Terrell Owens    12   37  ...  127.0        75.000000  226.000000
746        Edgerrin James    11   31  ...   11.0        73.000000  214.000000
1542      Matt Hasselbeck    15   38  ...    0.0        76.000000  118.933333
2158        Tony Gonzalez    15   37  ...  107.0        77.000000  251.000000
750           Eli Manning    10   32  ...    0.0        76.000000   87.200000
1752        Philip Rivers    10   32  ...    0.0        77.000000   91.200000
1605         Michael Vick    11   33  ...    0.0        73.000000   97.727273
2052          Steve Smith    18   34  ...   79.0        69.000000  185.000000
301         Carson Palmer    10   34  ...    0.0        77.000000   92.000000
23        Adrian Peterson    15   30  ...    5.0        63.333333  211.866667
1354          Kurt Warner    11   38  ...    0.0        74.000000  120.000000
2168            Tony Romo    10   33  ...    0.0        74.000000   87.600000

[20 rows x 22 columns]

In [ ]:
main_variables['Avg FP Per Game'] = main_variables['FantPt'] / main_variables['G']

In [ ]:
main_variables.sort_values(by='Avg FP Per Game',ascending=False).head(20)

Player  Year  ...      Weight  Avg FP Per Game
291            Cam Newton     3  ...  173.333333        20.708333
12          Aaron Rodgers     9  ...   74.333333        20.031915
1865       Robert Griffin     2  ...  110.000000        18.964286
1750       Peyton Manning    14  ...  115.000000        18.348214
77            Andrew Luck     2  ...  117.500000        17.781250
706            Drew Brees    13  ...   96.461538        17.741935
1564     Matthew Stafford     5  ...   92.000000        17.704918
554      Daunte Culpepper    11  ...  144.000000        17.523810
1689           Nick Foles     2  ...  122.000000        17.300000
2143            Tom Brady    14  ...  112.500000        17.113990
1828          Rich Gannon     6  ...  175.000000        17.067568
1921       Russell Wilson     2  ...  101.500000        17.062500
1760        Priest Holmes     8  ...  213.000000        16.788889
1367  LaDainian Tomlinson    11  ...  221.000000        16.770588
109          Arian Foster     5  ...  232.000000        16.745763
1522       Marshall Faulk     7  ...  211.000000        16.454545
690        Donovan McNabb    13  ...  129.230769        16.341317
1552            Matt Ryan     6  ...   74.666667        15.702128
1354          Kurt Warner    11  ...  120.000000        15.658537
81            Andy Dalton     3  ...  143.333333        15.583333

[20 rows x 23 columns]

In [ ]:
print ("Rows     : " , main_variables.shape[0])
print ("Columns  : " , main_variables.shape[1])
print ("\nFeatures : \n" , main_variables.columns.tolist())
print ("\nMissing values :  ", main_variables.isnull().sum().values.sum())
print ("\nUnique values :  \n",main_variables.nunique())

Rows     :  2302
Columns  :  23

Features : 
 ['Player', 'Year', 'Age', 'Tm', 'FantPt', 'G', 'GS', 'Cmp', 'Att', 'Yds', 'TD', 'Int', 'Att.1', 'Yds.1', 'Y/A', 'TD.1', 'Rec', 'Yds.2', 'Y/R', 'TD.2', 'Height (inches)', 'Weight', 'Avg FP Per Game']

Missing values :   1

Unique values :  
 Player             2302
Year                 17
Age                  23
Tm                   17
FantPt              627
G                   188
GS                  152
Cmp                 208
Att                 225
Yds                 287
TD                   98
Int                  87
Att.1               359
Yds.1               592
Y/A                 882
TD.1                 66
Rec                 352
Yds.2               965
Y/R                1290
TD.2                 69
Height (inches)      52
Weight              264
Avg FP Per Game    1380
dtype: int64


In [ ]:
main_variables.isnull().sum()

Player             0
Year               0
Age                0
Tm                 0
FantPt             0
G                  0
GS                 0
Cmp                0
Att                0
Yds                0
TD                 0
Int                0
Att.1              0
Yds.1              0
Y/A                0
TD.1               0
Rec                0
Yds.2              0
Y/R                0
TD.2               0
Height (inches)    0
Weight             0
Avg FP Per Game    1
dtype: int64

In [ ]:
main_variables['Avg FP Per Game'].replace(np.nan,0,inplace=True)

In [ ]:
main_variables.isnull().sum()

Player             0
Year               0
Age                0
Tm                 0
FantPt             0
G                  0
GS                 0
Cmp                0
Att                0
Yds                0
TD                 0
Int                0
Att.1              0
Yds.1              0
Y/A                0
TD.1               0
Rec                0
Yds.2              0
Y/R                0
TD.2               0
Height (inches)    0
Weight             0
Avg FP Per Game    0
dtype: int64